File: typescript/tests/personal_mental_board.test.ts
// personal_mental_board.test.ts
// Verifies the CircleAI.Personal.Mental port: mood logging + 7-day window and
// average (NaN when empty), journal entries (blank id throws, newest-first),
// and coping strategies by tag. The 7-day clock is injected for determinism.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryMentalHealthBoard,
  PersonalMentalDomainContext,
  Mood,
  moodLog,
  journalEntry,
  copingStrategy,
} from "../src/personal/mental/index";

const NOW = new Date("2026-07-10T12:00:00Z");
function board(): InMemoryMentalHealthBoard {
  const b = new InMemoryMentalHealthBoard();
  b.nowUtc = () => NOW;
  return b;
}

describe("Mood ordinals", () => {
  it("VeryLow=0 .. Great=4", () => {
    assert.equal(Mood.VeryLow, 0);
    assert.equal(Mood.Low, 1);
    assert.equal(Mood.Neutral, 2);
    assert.equal(Mood.Good, 3);
    assert.equal(Mood.Great, 4);
  });
});

describe("InMemoryMentalHealthBoard — mood", () => {
  it("Last7Days keeps entries within the window, sorted ascending", () => {
    const b = board();
    const withinA = new Date(NOW.getTime() - 6 * 24 * 3600 * 1000);
    const withinB = new Date(NOW.getTime() - 1 * 24 * 3600 * 1000);
    const old = new Date(NOW.getTime() - 8 * 24 * 3600 * 1000);
    b.logMood(moodLog(Mood.Good, withinB, null));
    b.logMood(moodLog(Mood.Low, withinA, "rough"));
    b.logMood(moodLog(Mood.Great, old, "too old"));
    assert.deepEqual(
      b.last7Days().map((m) => m.mood),
      [Mood.Low, Mood.Good],
    );
  });

  it("AvgMood7Day averages the ordinals; NaN when there are none", () => {
    const b = board();
    assert.ok(Number.isNaN(b.avgMood7Day()));
    b.logMood(moodLog(Mood.Neutral, new Date(NOW.getTime() - 3600 * 1000), null)); // 2
    b.logMood(moodLog(Mood.Great, new Date(NOW.getTime() - 7200 * 1000), null)); // 4
    assert.equal(b.avgMood7Day(), 3);
  });
});

describe("InMemoryMentalHealthBoard — journal", () => {
  it("entries list newest-first; blank id throws", () => {
    const b = board();
    b.addEntry(journalEntry("e1", "First", "…", new Date("2026-07-01T00:00:00Z")));
    b.addEntry(journalEntry("e2", "Second", "…", new Date("2026-07-05T00:00:00Z")));
    assert.deepEqual(
      b.entries.map((e) => e.entryId),
      ["e2", "e1"],
    );
    assert.throws(() => b.addEntry(journalEntry("   ", "X", "Y", new Date())), /EntryId required/);
  });
});

describe("InMemoryMentalHealthBoard — coping strategies", () => {
  it("StrategiesByTag matches case-insensitively; blank tag throws", () => {
    const b = board();
    b.registerStrategy(copingStrategy("s1", "Box Breathing", "…", ["Anxiety", "Grounding"]));
    b.registerStrategy(copingStrategy("s2", "5-4-3-2-1", "…", ["grounding"]));
    b.registerStrategy(copingStrategy("s3", "Gratitude", "…", ["Mood"]));
    assert.deepEqual(
      b.strategiesByTag("GROUNDING").map((s) => s.strategyId).sort(),
      ["s1", "s2"],
    );
    assert.throws(() => b.strategiesByTag(" "), /tag required/);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(PersonalMentalDomainContext.systemPromptSnippet.includes("[DOMAIN: Personal.Mental]"));
    assert.deepEqual(PersonalMentalDomainContext.complianceFlags, [
      "POPIA",
      "Mental_Health_Care_Act_17_2002",
      "Not_Therapy",
      "Crisis_Protocol",
    ]);
    assert.deepEqual(PersonalMentalDomainContext.suggestedTools, ["journal", "breathing_tools", "mood_tracker", "web_search"]);
  });
});

File: go/tests/personal_mental_board_test.go
// personal_mental_board_test.go
//
// Verifies the CircleAI.Personal.Mental port (personal_mental_board.go): Mood
// ordinals/names, mood logging + 7-day window + average (NaN when empty),
// journal add (blank-id error) + newest-first ordering, and coping-strategy tag
// lookup (case-insensitive, blank-tag error).

package circleai_test

import (
	"math"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestPersonalMental_MoodOrdinals(t *testing.T) {
	if circleai.MoodVeryLow != 0 || circleai.MoodGreat != 4 {
		t.Fatalf("ordinals: verylow=%d great=%d", circleai.MoodVeryLow, circleai.MoodGreat)
	}
	if circleai.MoodNeutral.String() != "Neutral" || circleai.MoodGreat.String() != "Great" {
		t.Fatalf("names wrong: %s / %s", circleai.MoodNeutral, circleai.MoodGreat)
	}
}

func TestPersonalMental_MoodWindowAndAverage(t *testing.T) {
	b := circleai.NewInMemoryMentalHealthBoard()
	// Empty window -> NaN.
	if !math.IsNaN(b.AvgMood7Day()) {
		t.Fatalf("empty avg should be NaN")
	}
	now := time.Now().UTC()
	b.LogMood(circleai.MoodLog{Mood: circleai.MoodGood, AtUtc: now.Add(-1 * time.Hour)})    // 3
	b.LogMood(circleai.MoodLog{Mood: circleai.MoodNeutral, AtUtc: now.Add(-2 * time.Hour)}) // 2
	b.LogMood(circleai.MoodLog{Mood: circleai.MoodGreat, AtUtc: now.Add(-3 * time.Hour)})   // 4
	// Older than 7 days -> excluded from window and average.
	b.LogMood(circleai.MoodLog{Mood: circleai.MoodVeryLow, AtUtc: now.Add(-8 * 24 * time.Hour)})

	last7 := b.Last7Days()
	if len(last7) != 3 {
		t.Fatalf("7-day window = %d, want 3", len(last7))
	}
	// Ascending by time: Great(-3h), Neutral(-2h), Good(-1h).
	if last7[0].Mood != circleai.MoodGreat || last7[2].Mood != circleai.MoodGood {
		t.Fatalf("window order failed: %+v", last7)
	}
	// Average of 3,2,4 = 3.0.
	if avg := b.AvgMood7Day(); math.Abs(avg-3.0) > 1e-9 {
		t.Fatalf("avg = %v, want 3.0", avg)
	}
}

func TestPersonalMental_JournalEntries(t *testing.T) {
	b := circleai.NewInMemoryMentalHealthBoard()
	base := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	if err := b.AddEntry(circleai.JournalEntry{EntryId: "e1", Title: "Day 1", Body: "...", AtUtc: base}); err != nil {
		t.Fatalf("add entry: %v", err)
	}
	_ = b.AddEntry(circleai.JournalEntry{EntryId: "e2", Title: "Day 3", Body: "...", AtUtc: base.Add(48 * time.Hour)})
	_ = b.AddEntry(circleai.JournalEntry{EntryId: "e3", Title: "Day 2", Body: "...", AtUtc: base.Add(24 * time.Hour)})

	entries := b.Entries()
	if len(entries) != 3 || entries[0].EntryId != "e2" || entries[1].EntryId != "e3" || entries[2].EntryId != "e1" {
		t.Fatalf("entries newest-first failed: %+v", entries)
	}
	if err := b.AddEntry(circleai.JournalEntry{EntryId: "  ", Title: "blank"}); err == nil {
		t.Fatalf("blank entry id must error")
	}
}

func TestPersonalMental_StrategiesByTag(t *testing.T) {
	b := circleai.NewInMemoryMentalHealthBoard()
	b.RegisterStrategy(circleai.CopingStrategy{StrategyId: "s1", Title: "Box Breathing", Description: "...", Tags: []string{"Anxiety", "Breathing"}})
	b.RegisterStrategy(circleai.CopingStrategy{StrategyId: "s2", Title: "Walk", Description: "...", Tags: []string{"Exercise"}})
	b.RegisterStrategy(circleai.CopingStrategy{StrategyId: "s3", Title: "Grounding", Description: "...", Tags: []string{"anxiety"}})

	hits, err := b.StrategiesByTag("ANXIETY")
	if err != nil {
		t.Fatalf("by tag: %v", err)
	}
	if len(hits) != 2 || hits[0].StrategyId != "s1" || hits[1].StrategyId != "s3" {
		t.Fatalf("case-insensitive tag match failed: %+v", hits)
	}
	if _, err := b.StrategiesByTag(""); err == nil {
		t.Fatalf("blank tag must error")
	}
}

func TestPersonalMental_StrategyTagsCopied(t *testing.T) {
	b := circleai.NewInMemoryMentalHealthBoard()
	tags := []string{"Anxiety"}
	b.RegisterStrategy(circleai.CopingStrategy{StrategyId: "s1", Title: "X", Tags: tags})
	tags[0] = "MUTATED"
	hits, _ := b.StrategiesByTag("Anxiety")
	if len(hits) != 1 {
		t.Fatalf("strategy tags not defensively copied")
	}
}
